### ISSUE TITLE
Test recipes using the Mock server may have intermittent failures

### ISSUE THREAD

**Issue by levitte** [openssl/openssl#14694]
I randomly get this sort of test failure, usually when I've more than one test happening at the same time:

``` console
Warning: certificate from 'trusted.crt' with subject '/O=openssl_cmp' is not a CA cert
# cmp_main:apps/cmp.c:2578:CMP info: using section(s) 'Mock connection' of OpenSSL configuration file '../Mock/test.cnf'
# opt_str:apps/cmp.c:2177:CMP warning: argument of -proxy option is empty string, resetting option
# setup_client_ctx:apps/cmp.c:1894:CMP info: will contact http://127.0.0.1:1700/pkix/
# CMP info: sending IR
# CMP info: received IP
# CMP info: sending CERTCONF
# CMP info: received ERROR
# CMP error: transactionid unmatched:
../../../../util/wrap.pl ../../../../apps/openssl cmp -config ../Mock/test.cnf -section 'Mock connection' -certout ../../../../test-runs/test_cmp_http/test.cert.pem -proxy '' -no_proxy 127.0.0.1 -server '127.0.0.1:1700/ignored' -path pkix/ => 1
    not ok 17 - path overrides -server path
# ------------------------------------------------------------------------------
    # Looks like you failed 1 test of 31.
not ok 1 - CMP app CLI Mock connection
# ------------------------------------------------------------------------------
Warning: certificate from 'trusted.crt' with subject '/O=openssl_cmp' is not a CA cert
# cmp_main:apps/cmp.c:2578:CMP info: using section(s) 'Mock verification' of OpenSSL configuration file '../Mock/test.cnf'
# opt_str:apps/cmp.c:2177:CMP warning: argument of -proxy option is empty string, resetting option
# setup_client_ctx:apps/cmp.c:1894:CMP info: will contact http://127.0.0.1:1700/pkix/
# CMP info: sending IR
# CMP info: received IP
# CMP info: sending CERTCONF
# CMP info: received ERROR
# CMP error: transactionid unmatched:
../../../../util/wrap.pl ../../../../apps/openssl cmp -config ../Mock/test.cnf -section 'Mock verification' -certout ../../../../test-runs/test_cmp_http/test.cert.pem -proxy '' -no_proxy 127.0.0.1 -recipient /O=openssl_cmp -trusted trusted.crt -unprotected_errors => 1
    not ok 1 - default test
# ------------------------------------------------------------------------------
    # Looks like you failed 1 test of 40.
not ok 2 - CMP app CLI Mock verification
# ------------------------------------------------------------------------------
Warning: certificate from 'trusted.crt' with subject '/O=openssl_cmp' is not a CA cert
Invalid password argument, missing ':' within the first 5 chars
Error getting password for private key for CMP client certificate
Trying plain input string (better precede with 'pass:')
Invalid password argument, missing ':' within the first 5 chars
Error getting password for CMP client certificate (optionally with chain)
Trying plain input string (better precede with 'pass:')
# cmp_main:apps/cmp.c:2578:CMP info: using section(s) 'Mock credentials' of OpenSSL configuration file '../Mock/test.cnf'
# opt_str:apps/cmp.c:2177:CMP warning: argument of -proxy option is empty string, resetting option
# setup_client_ctx:apps/cmp.c:1894:CMP info: will contact http://127.0.0.1:1700/pkix/
# CMP info: sending IR
# CMP info: received IP
# CMP info: sending CERTCONF
# CMP info: received ERROR
# CMP warning: no secret available for verifying PBM-based CMP message protection
# CMP error: received error:PKIStatus: rejection; PKIFailureInfo: badRequest; errorCode: 133; errorDetails: unexpected pkibody
../../../../util/wrap.pl ../../../../apps/openssl cmp -config ../Mock/test.cnf -section 'Mock credentials' -certout ../../../../test-runs/test_cmp_http/test.cert.pem -proxy '' -no_proxy 127.0.0.1 -cert signer.crt -key signer.p12 -keypass 12345 => 1
    not ok 13 - keypass no prefix
# ------------------------------------------------------------------------------
    # Looks like you failed 1 test of 38.
not ok 3 - CMP app CLI Mock credentials
# ------------------------------------------------------------------------------
Warning: certificate from 'trusted.crt' with subject '/O=openssl_cmp' is not a CA cert
# cmp_main:apps/cmp.c:2578:CMP info: using section(s) 'Mock commands' of OpenSSL configuration file '../Mock/test.cnf'
# opt_str:apps/cmp.c:2177:CMP warning: argument of -proxy option is empty string, resetting option
# setup_request_ctx:apps/cmp.c:1485:CMP warning: -subject option is ignored for commands other than 'ir', 'cr', and 'kur' since -ref or -cert is given
# setup_request_ctx:apps/cmp.c:1497:CMP warning: -policies option is ignored for commands other than 'ir', 'cr', and 'kur'
# setup_request_ctx:apps/cmp.c:1535:CMP warning: -newkey option is ignored for 'p10cr' and 'rr' commands
# setup_client_ctx:apps/cmp.c:1894:CMP info: will contact http://127.0.0.1:1700/pkix/
# CMP info: sending RR
# CMP error: connect timeout:
# CMP error: passed a null parameter:
# CMP error: total timeout:request sent: RR, expected response: RP
../../../../util/wrap.pl ../../../../apps/openssl cmp -config ../Mock/test.cnf -section 'Mock commands' -certout ../../../../test-runs/test_cmp_http/test.cert.pem -proxy '' -no_proxy 127.0.0.1 -cmd rr -revreason 0 -csr csr.pem => 1
    not ok 30 - --- use csr for revocation ----
# ------------------------------------------------------------------------------
    #   Failed test '--- use csr for revocation ----'
    #   at test/recipes/80-test_cmp_http.t line 145.
Warning: certificate from 'trusted.crt' with subject '/O=openssl_cmp' is not a CA cert
# cmp_main:apps/cmp.c:2578:CMP info: using section(s) 'Mock commands' of OpenSSL configuration file '../Mock/test.cnf'
# opt_str:apps/cmp.c:2177:CMP warning: argument of -proxy option is empty string, resetting option
# setup_client_ctx:apps/cmp.c:1894:CMP info: will contact http://127.0.0.1:1700/pkix/
# CMP info: sending CR
# CMP error: connect timeout:
# CMP error: passed a null parameter:
# CMP error: total timeout:request sent: CR, expected response: CP
../../../../util/wrap.pl ../../../../apps/openssl cmp -config ../Mock/test.cnf -section 'Mock commands' -certout ../../../../test-runs/test_cmp_http/test.cert.pem -proxy '' -no_proxy 127.0.0.1 -cmd cr => 1
    not ok 31 - --- get certificate for revocation ----
# ------------------------------------------------------------------------------
    # Looks like you failed 2 tests of 48.
not ok 4 - CMP app CLI Mock commands
# ------------------------------------------------------------------------------
Warning: certificate from 'trusted.crt' with subject '/O=openssl_cmp' is not a CA cert
# cmp_main:apps/cmp.c:2578:CMP info: using section(s) 'Mock enrollment' of OpenSSL configuration file '../Mock/test.cnf'
# opt_str:apps/cmp.c:2177:CMP warning: argument of -proxy option is empty string, resetting option
# setup_client_ctx:apps/cmp.c:1894:CMP info: will contact http://127.0.0.1:1700/pkix/
# CMP info: sending IR
# CMP info: received IP
# CMP info: sending CERTCONF
# CMP info: received ERROR
# CMP error: transactionid unmatched:
../../../../util/wrap.pl ../../../../apps/openssl cmp -config ../Mock/test.cnf -section 'Mock enrollment' -certout ../../../../test-runs/test_cmp_http/test.cert.pem -proxy '' -no_proxy 127.0.0.1 -cmd ir -newkey new.key -newkeypass 'pass:' -sans critical -certout ../../../../test-runs/test_cmp_http/test.certout_sans_critical.pem -out_trusted root.crt => 1
    not ok 35 - sans critical
# ------------------------------------------------------------------------------
    # Looks like you failed 1 test of 92.
not ok 5 - CMP app CLI Mock enrollment
# ------------------------------------------------------------------------------
# 
#   Failed test 'CMP app CLI Mock enrollment
# '
#   at /var/lib/buildbot/workers/openssl-experimental/fips_build_test_debian-x86_64_default_clang_/build/util/perl/OpenSSL/Test.pm line 1335.
kill: (12463): No such process
# Looks like you failed 5 tests of 5.80-test_cmp_http.t ................. 
Dubious, test returned 5 (wstat 1280, 0x500)
Failed 5/5 subtests 

...

Test Summary Report
-------------------
80-test_cmp_http.t               (Wstat: 1280 Tests: 5 Failed: 5)
  Failed tests:  1-5
  Non-zero exit status: 5
Files=234, Tests=3294, 3363 wallclock secs (18.28 usr  1.19 sys + 721.02 cusr 73.59 csys = 814.08 CPU)
Result: FAIL
make[1]: *** [Makefile:3337: _tests] Error 1
make[1]: Leaving directory '/var/lib/buildbot/workers/openssl-experimental/fips_build_test_debian-x86_64_default_clang_/build'
make: *** [Makefile:3334: tests] Error 2
program finished with exit code 2
```

Why this happens is quite clear:

https://github.com/openssl/openssl/blob/8c63532002fdab11b437bc8d68012c2b05cf00ea/test/recipes/80-test_cmp_http_data/Mock/test.cnf#L20

Two tests running at the same time, and specifically running this test recipe simultaneously will obviously compete for resources.

We did resolve this very same issue, a few years ago, with the `util/perl/TLSProxy` code.  Can we do it again?

**Comment by slontis** [comment]
@DDvO

**Comment by DDvO** [comment]
> server_port = 1700 

> Two tests running at the same time, and specifically running this test recipe simultaneously will obviously compete for resources.

True!

> We did resolve this very same issue, a few years ago, with the util/perl/TLSProxy code. Can we do it again?

Hmm, the TLSProxy code looks like a very heavyweight solution. 
Can anyone who was involved in it (apparently, mostly @mattcaswell) 
refactor it in a way such that it is usable also for tests like those in `80-test_cmp_http.t`?

**Comment by paulidale** [comment]
#14966 likely relates to this.  With the AFL it seems to be highly repeatable.  I'm not sure it is actually fuzzing anything though.

**Comment by DDvO** [comment]
The intermittent test failures have a different source than #14966 - see my comment there.

**Comment by levitte** [comment]
I didn't mean to say that we should use TLSProxy with the Mock server.  All I want is a mechanism for it to listen on a *random port* rather than a static one.

What TLSProxy does is to run `openssl s_server` with `-accept localhost:0`.  and then parses the output for the line saying `ACCEPT:`, which specifies the address and actual port.  The Mock server could be made to do something similar.

**Comment by DDvO** [comment]
Good thought - i'm going to do that now.

**Comment by DDvO** [comment]
Fixed by #15279.
